### ISSUE TITLE
Bridge(Handler) and Thing(Handler) relationship

### ISSUE THREAD

**Issue by maggu2810** [eclipse/smarthome#1479]
If there is a relationship between the bridge and the thing, how could this implemented in the handler correctly?
Let's assume a Bridge does need to know if a Thing has been added or removed e.g. because there has to be some resources reserved and released from the bridge. If the thing ever received such a resource, it uses this one all the time until it is removed (e.g. an ID). The BridgeHandler has to manage the pool of such IDs, so it does need to know which one has been already assigned and which one are free to use.

Is this supported by the current architecture?
Are the "bridgeHandler..." methods of the Thing the methods of be used?
- in initialized the thing handler checks if the thing property contains already an ID
- it calls getBridge and if it returns not null it calls bridge.getNewId()
- if getBridge returns null it waits for bridgeHandlerInitialized and calls bridge.getNewId()
- if updateStatus of the thing is set to REMOVING it calls bridge.releaseId(id)
- if the bridge is changed to another one I need to call bridgeOld.releaseId(id), bridgeNew.getNewId()

@sbussweiler I assume you can help?

**Comment by maggu2810** [comment]
The bridge or better the bridge handler of the bridge manages a set of IDs.
Every thing that is part of the bridge gets one ID.
If the thing is deleted the bridge does need to know that the ID is free and can be used by another thing.

So the bridge does need to know if a thing is no longer using that bridge (the thing is removed or another bridge is assigned).

Any hints how to implement this?

**Comment by sbussweiler** [comment]
You want to know if a Bridge-Child-Thing has been added or removed in order to allocate and deallocate resources which are managed by the Bridge?

> 1. in initialized the thing handler checks if the thing property contains already an ID
2. it calls getBridge and if it returns not null it calls bridge.getNewId()
3. if getBridge returns null it waits for bridgeHandlerInitialized and calls bridge.getNewId()
4. if updateStatus of the thing is set to REMOVING it calls bridge.releaseId(id)
5. if the bridge is changed to another one I need to call bridgeOld.releaseId(id), bridgeNew.getNewId()

So far this sounds good to me with two exceptions:

1. ThingHandler.bridgeHandlerInitialized() is deprecated. Please use ThingHandler.bridgeStatusChanged()

2. I would use ThingHandler.dispose() to release resp. deallocate such resources. This method is called if a thing has been removed AND if you restart your bundle/system.

**Comment by maggu2810** [comment]
Can you confirm that "ThingHandler.bridgeStatusChanged()" called twice (for the old and for the new) if the bridge is changed of a thing?

**Comment by sbussweiler** [comment]
If the status of the old/new bridge has changed to ONLINE/OFFLINE - yes.

**Comment by maggu2810** [comment]
Okay, thanks again.
I will try that methods and see if it is working (but I trust you :wink: ).

**Comment by maggu2810** [comment]
> If the status of the old/new bridge has changed to ONLINE/OFFLINE - yes.

I have thought of another scenario:
* there are two bridges and one thing
* bridge_1 is online, bridge_2 is online
* the thing is using bridge_1
* the thing has reserved one ID of bridge_1
* the thing is moved from bridge_1 to bridge_2

The ThingHandler does need to be notified about such a change.
It has to be release the ID of bridge_1 and request one of bridge_2.

**Comment by kaikreuzer** [comment]
I today talked to @chrisschauer who (rightly?) complained about the complexity of all this interaction sequence. He wondered, why the initialization is not simply postponed until the BridgeHandler is initialized (and disposed again before the BridgeHandler is unregistered). I couldn't give a good answer to that, but it might be worth to spend some thoughts whether this is a possibility which would simplify things for developers.

**Comment by chrisschauer** [comment]
Thanks @kaikreuzer for starting the discussion. I think the following topic could be improved for binding development.

**Initialization mechanism**
ESH includes a bridge <-> thing hierarchy which is defined in the thing.xml. But the initialization mechanisum does not walk through the hierarchy and initializes a bridge first. Due to this it's possible a thing is initialized befor the associated bridge is initialized. Some bindings use e.g a property for configuring a update interval (most in the bridge). An immediately started background job will crash, because the bridge (with the update intervall configuration) isn't initialized yet. 

My wish: ESH should call the handler factory in correct order within the thing-hierarchy like walking through a tree from root to leafs.

**Comment by maggu2810** [comment]
The XML currently provides a bridge-ref for things.
Does it make sense to add a boolean mandatory attribute, so the framework could prevent creation of things without a bridge of mandatory is true?

**Comment by chrisschauer** [comment]
Are there things which can be initialized without their bridge?
I would do it in general without extra configuration. If a supported bridge is defined, i would always initialize it first?

I thought a thing (with defined supported bridge type) always have their bridge regardless of the current state in thing lifecycle.

**Comment by maggu2810** [comment]
For my current use case I am fine if the bridge is always present.
But I don't know if there are binding that implements things that bridge is optional.
E.g. if a bridge is defined use it, if not use a direct connection.

**Comment by chrisschauer** [comment]
It also works for all of our bindings. In the case a bridge is optional, i would prefer to mark a bridge of optional like:

```xml
  <supported-bridge-type-refs optional="true">
            <bridge-type-ref id="bridge_type_foo" />
  </supported-bridge-type-refs>
```

**Comment by maggu2810** [comment]
I am fine with you suggestion (personally).

But there are not only yours and mine bindings.
I don't know if there are other private bindings that gets broken after an incompatible change (if we change the default from "is optional" to "is not optional").
For the listed attribute the decision has been made this way to keep the logic as before.

But that is a general decision, does the ESH SNAPSHOTs needs to be as compatible as possible or could we break bindings (so the very public API and not internal stuff).

**Comment by kaikreuzer** [comment]
As we are still on a <1.0 version and the binding APIs have never been declared final, I support breaking changes, if we notice that the current setup is overly complex and can be heavily simplified.

Adding an optional flag and having to support two different lifecycles makes things more complicated and error-prone, so I would only go this way if we really see a need for both.

**Comment by maggu2810** [comment]
> Adding an optional flag and having to support two different lifecycles makes things more complicated and error-prone, so I would only go this way if we really see a need for both.

So, I am for all my bindings fine to not introduce a flag and handle a supported bridge type reference this way, that a bridge is required.

I am also fine with the lifecycle handling change a bridge is always initialized first.

For the notifications / callbacks:
I don't know if it is better
* to inform the thing handler that the assigned bridge has been changed, so the thing handler could call functions to request / release resources
* to inform the bridge handler has a thing has been assigned / de-assigned, so the bridge handler could call functions to request / release resources

I assume that normally the bridge manages resources and so the bridge handler should be informed. So a bug in the thing handler corruption does not eat (and never release) resources of the bridge.

**Comment by sbussweiler** [comment]
I like the idea that a Bridge would always be initialized first. That makes the Bridge <-> Thing lifecycle easier and transparent. I agree also with @kaikreuzer, supporting an optional flag and supporting two different lifecycle makes it much more complicated. I would always initialize a Bridge first. But as mentioned several times in this issue; I have no idea if there are any bindings with Things and optional Bridges.

Regarding the notifications/callbacks: for me the Bridge is the "central node", so the bridge handler should be notified about assigned / de-assigned  Things. Anyway, we can keep handler methods like ThingHandler.bridgeStatusChanged().

**Comment by maggu2810** [comment]
Are this the changes (that needs to be done) we agreed?

* a bridge is initialized before the things that are using the bridge are initialized
* a bridge is disposed after the all the things that are using the bridge are disposed
* a bridge is mandatory (not optional) if a bridge is referenced in the thing type
* the bridge handler should be notified if a thing is assigned / de-assigned

**Comment by chrisschauer** [comment]
These are exactly the points i expect from ESH. The notification if athing is assigned / de-assigned is nice to have. Thx

**Comment by kaikreuzer** [comment]
Ok, so now we only need a volunteer to implement this change. Afair, @sbussweiler is on leave for a while, so I assume we cannot count on him?

**Comment by sbussweiler** [comment]
If there is no other volunteer (?), I can try my best. But I can not promise when I can deliver it ;)

**Comment by maggu2810** [comment]
I am too busy ATM. So it would be great if you can do so.

**Comment by marcelrv** [comment]
+1 for simplifying this piece

**Comment by lsafelix75** [comment]
Please implement this feature. My current code has a lot of checking on thing to make sure that bridge has already been created.

Im not sure whether this is a right chatroom. How does this affect Paper UI? As far as i know whenever we convert an inbox item to thing, bridgehandler is not assigned until you manually do it on Things setting. With above suggestion, does mean a thing will only initialized once a bridge is assigned?

**Comment by kaikreuzer** [comment]
> With above suggestion, does mean a thing will only initialized once a bridge is assigned?

Yes, this would be the result. Do you expect any side effects from this? Is there any reason why such a Thing (without an assigned bridge) should already have a handler instance?

**Comment by maggu2810** [comment]
Hi @lsafelix75, I think we already agreed that this should be changed but to the lack of time nobody has implemented this. See https://github.com/eclipse/smarthome/issues/1479#issuecomment-221818491. If you can create a PR for, we all will be happy.

**Comment by marcelrv** [comment]
@sbussweiler 
Seems bridgeStatusChanged() is called for each status update from the bridge.
Shouldn't that be only when the status has actually changed? (otherwise it should be called bridgeStatusUpdated() ...)

Also seems like it is not called when the bridge is removed. I thought the intention was to depreciate bridgeHandlerDisposed(), in that case I think we should call bridgeStatusChanged() also when state is in removing/removed

**Comment by sbussweiler** [comment]
> Seems bridgeStatusChanged() is called for each status update from the bridge.
Shouldn't that be only when the status has actually changed? (otherwise it should be called bridgeStatusUpdated() ...)

Really? I think there is only a call if the status has been changed. See ThingManager [here](https://github.com/eclipse/smarthome/blob/master/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/internal/ThingManager.java#L203).

> Also seems like it is not called when the bridge is removed. I thought the intention was to depreciate bridgeHandlerDisposed(), in that case I think we should call bridgeStatusChanged() also when state is in removing/removed

That's correct. Currently bridgeStatusChanged() is only called for status changes to ONLINE/OFFLINE. I think we missed that in our discussion (https://github.com/eclipse/smarthome/issues/1075#issuecomment-205710734). 

I hope that I will find some time to start with the new bridge/thing lifecycle implementation in the next weeks.

**Comment by kgoderis** [comment]
@sbussweiler ( @kaikreuzer ) I just noticed that when shutting down the runtime, and when the Bridge happen to be disposed before the child Things, the method in the child ThingHandler     `public void bridgeHandlerDisposed(ThingHandler bridgeHandler, Bridge bridge)`
is called with bridgeHandler being null 

In the new lifecycle, would it not be more logical to have the Bridges dispose *after* their child Things? The children typically depend on the bridges for connectivity or alike, and maybe they need to do one last action or whatever using that bridge before they can be destroyed.

**Comment by chrisschauer** [comment]
Hi @kgoderis 

this is exactly the requirement which was previously discussed: See the summary in: https://github.com/eclipse/smarthome/issues/1479#issuecomment-221282330

Its also my wish to have an hierarchically correct orde

**Comment by marcelrv** [comment]
> Really? I think there is only a call if the status has been changed. See ThingManager here.
think you are right. Guess I was seeing some artifact from the recent startup issues that was causing it not to work well. Now that is fixed I don't see it happening.

**Comment by kgoderis** [comment]
@chrisschauer Tx, I should take more time to read all comments top to bottom ;-)

**Comment by kaikreuzer** [comment]
@sbussweiler FTR: There seem to be MANY issues with most bindings regarding the bridge/thing startup lifecycle, see [this discussion](https://community.openhab.org/t/binding-not-starting-consistently-since-6-19-build/11741/93?u=kai). I therefore think that this has a very high priority and really must be addressed. See also https://github.com/eclipse/smarthome/issues/1869.

**Comment by sbussweiler** [comment]
@kaikreuzer I'm trying to start this week with the implementation ;-)

**Comment by sbussweiler** [comment]
Hello @all,

I provided a first implementation of the Bridge-Thing life cycle refactoring. Sorry, that it took soo long, but I was really busy.

You can find the current implementation [here](https://github.com/qivicon/smarthome/tree/thing-bridge-lifecycle-refactoring). Please note, this is not the final state in order to create a pull request. There is an ongoing minor API discussion, missing JavaDoc and no documentation right now. The plan is to (hopefully) create a PR next week. Until then it would be great if some of you can check the implementation as follows: are all your requirements to 100% fulfilled, does the implementation work as expected with your bindings, etc. 

It would be glad to get some feedback because it's a major life cycle refactoring. Especially from @chrisschauer and @maggu2810, you mainly requested it ;-)


**_First, some - important to know - issues since there is no updated documentation right now:_**

- I removed the OSGi service registration of thing handlers. All handlers are managed by the `ThingManager`. There is no need to have a handler registered as an OSGi service. The service registration was overhead and the life cycle is simplified a lot. For more information see discussion #955.

- I introduced a new interface called `BridgeHandler` (yes, there was no BridgeHandler interface before) which provides two methods to inform about child handler initialization and disposal. There are no compilation errors as long as all bridge handlers are implementing the base class `BaseBridgeHandler` (which is recommended). The `BaseBridgeHandler` provides an empty default implementation. 

- The methods `ThingHandler.bridgeHandlerInitialized()` and `ThingHandler.bridgeHandlerDisposed()` have been removed, which breaks the API. In my opinion the methods are not longer needed. 

- The framework (`ThingManager`) was responsible to set the status of child things to 
  -  `OFFLINE` with detail `BRIDGE_OFFLINE` if a bridge goes `OFFLINE`
   -  `OFFLINE` with detail `NONE` if a bridge goes `ONLINE` again

 Due to the fact that a binding is responsible to set the ONLINE/OFFLINE status, this implementation has been moved to `BaseThingHandler.bridgeStatusChanged(ThingStatusInfo bridgeStatusInfo)`. Means, all handler implementations (which are implementing this method) must make a `super.bridgeStatusChanged(bridgeStatusInfo)` call in order to fulfill the contract. 


**_The following requirements (collected in this issue) should be fulfilled:_**

> a bridge is initialized before the things that are using the bridge are initialized

If a child thing is added and the bridge has not been initialized, the child thing handler will be ignored. If the bridge has been added and the bridge handler is initialized (thing status is ONLINE resp. OFFLINE) all child handlers of the available things are initialized.

> a bridge is disposed after all the things that are using the bridge are disposed

If a bridge should be disposed, first all child thing handlers are disposed. `ThingHandler.dispose()` must return immediately.

> a bridge is mandatory (not optional) if a bridge is referenced in the thing type

There is no optional configuration in order to avoid this, the bridge is always mandatory.

> the bridge handler should be notified if a thing is assigned / de-assigned

A `BridgeHandler` is notified about the thing initialization via `BridgeHandler.childHandlerInitialized(ThingHandler, Thing)`. About the disposal of a thing a bridge handler will get informed by the method `BridgeHandler.childHandlerDisposed(ThingHandler, Thing)`.


> a thing handler is informed about status changes of the bridge to ONLINE/OFFLINE

This behavior is ensured by the known method `ThingHandler.bridgeStatusChanged(ThingStatusInfo bridgeStatusInfo)`. Please note that handler implementation must make a super-call.

**Comment by kgoderis** [comment]
@sbussweiler Just some thought : 

> a thing handler is informed about status changes of the bridge to ONLINE/OFFLINE

should not happen if

> If a child thing is added and the bridge has not been initialized, the child thing handler will be ignored. If the bridge has been added and the bridge handler is initialized (thing status is ONLINE resp. OFFLINE) all child handlers of the available things are initialised.

is the case/happening, no? e.g. avoid that a ThingHandler receives a bridgeStatusChanged(ThingStatusInfo bridgeStatusInfo) before it is initialised, as the initialisation of the bridge itself will trigger a change of its status, and hence the call to the child ThingHandler

**Comment by kgoderis** [comment]
> ThingHandler.dispose() must return immediately.

What do you mean exactly? that a dispose() should be quick/fast? but what with the use case that a child ThingHandler needs to communicate with the physical device to unregister it from a network or alike?

**Comment by kgoderis** [comment]
Just being the devil's advocate: does all of this still make sense if there is a cascade of Bridges? (not that I know of any binding having this, but in theory - at least a long time ago - it was theoretically possible to do that)

**Comment by kgoderis** [comment]
> There is no need to have a handler registered as an OSGi service

mmh... I guess you mean for the purpose of simply tracking them, right? In some occasions handlers need access to one or the other Registry. In order to simplify things, I often have that binding's ThingHandlerFactory (since it is a singleton anyways) bind the OSGi services, and pass them on as arguments to the ThingHandler constructors. If the ThingManager is the "disc jokey" for all the ThingHandlers, then maybe there is an opportunity to generalise this and have the ThingManager expose/provide some of these kids of functionalities to the ThingHandlers, thereby simplifying some aspects of binding development?

**Comment by maggu2810** [comment]
@sbussweiler I will try to test your changes, but it takes some time.

**Comment by sbussweiler** [comment]
@kgoderis, thanks for your feedback!

> is the case/happening, no? e.g. avoid that a ThingHandler receives a bridgeStatusChanged(ThingStatusInfo bridgeStatusInfo) before it is initialised, as the initialisation of the bridge itself will trigger a change of its status, and hence the call to the child ThingHandler. Maybe bridgeStatusChanged() needs an additional parameter to indicate if the status changed during or due to initialisation?

`ThingHandler.bridgeStatusChanged()` is never called before a child thing has been initialized. The first bridge status change to ONLINE/OFFLINE is initiated by the bridge initialization itself. In that case `ThingHandler.bridgeStatusChanged()` is _not_ called (because it is not really a change), but `ThingHandler.initialize()` of all child things. I refactored this behavior this morning, please pull the latest changes.

> What do you mean exactly? that a dispose() should be quick/fast? but what with the use case that a child ThingHandler needs to communicate with the physical device to unregister it from a network or alike?

The framework expects `dispose()` to be non-blocking and return quickly. For longer running disposals, the implementation has to take care of scheduling a separate job

> Just being the devil's advocate: does all of this still make sense if there is a cascade of Bridges? (not that I know of any binding having this, but in theory - at least a long time ago - it was theoretically possible to do that)

In theory it should still be possible  ;-)

> mmh... I guess you mean for the purpose of simply tracking them, right? In some occasions handlers need access to one or the other Registry. In order to simplify things, I often have that binding's ThingHandlerFactory (since it is a singleton anyways) bind the OSGi services, and pass them on as arguments to the ThingHandler constructors. If the ThingManager is the "disc jokey" for all the ThingHandlers, then maybe there is an opportunity to generalise this and have the ThingManager expose/provide some of these kids of functionalities to the ThingHandlers, thereby simplifying some aspects of binding development?

We can extend the `ThingHandlerCallback` in order to provide common used services (e.g. `callback.getThingTypeRegistry()`) to our thing handlers. Means, we have to collect important and common used services. But I think this must not be part of this bridge/thing life cycle refactoring. We can deliver something in an upcoming pull request.

**Comment by maggu2810** [comment]
@sbussweiler If I merge your code in the current master branch and start a build, I run into a build failure.
See build.stdout.txt...

[build.stderr.txt](https://github.com/eclipse/smarthome/files/442691/build.stderr.txt)
[build.stdout.txt](https://github.com/eclipse/smarthome/files/442692/build.stdout.txt)

**Comment by sbussweiler** [comment]
@maggu2810 very strange... On my machine the build is green. I think you have the latest version from github and tried to run it more then once? Maybe a timing issue?

**Comment by maggu2810** [comment]
I have running other tasks on my machine in parallel, so perhaps a timing problem. But at least it shows that it is not stable or the test is error-prone.
I used the recent master of yesterday and merged your branch...
I can skip the tests, but if there is still some timing related issue, perhaps we should fix that first. ? WDYT?

**Comment by sjka** [comment]
I also tried to reproduce it but couldn't. However, I ran into this one in ~12% of the iterations:

```
ThingManager calls bridgeStatusChanged on ThingHandler correctly(org.eclipse.smarthome.core.thing.internal.ThingManagerOSGiTest)  Time elapsed: 10.1 sec  <<< FAILURE!
java.lang.AssertionError: 
Expected: is <ONLINE>
     but: was <UNINITIALIZED>
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.junit.Assert.assertThat(Assert.java:865)
	at org.junit.Assert.assertThat(Assert.java:832)
	at org.junit.Assert$assertThat.callStatic(Unknown Source)
	at org.eclipse.smarthome.core.thing.internal.ThingManagerOSGiTest$_ThingManager_calls_bridgeStatusChanged_on_ThingHandler_correctly_closure239.doCall(ThingManagerOSGiTest.groovy:1057)
	at org.eclipse.smarthome.core.thing.internal.ThingManagerOSGiTest$_ThingManager_calls_bridgeStatusChanged_on_ThingHandler_correctly_closure239.doCall(ThingManagerOSGiTest.groovy)
	at sun.reflect.GeneratedMethodAccessor66.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:90)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:324)
	at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:278)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1016)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:39)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:45)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:108)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:112)
	at org.eclipse.smarthome.test.OSGiTest.waitForAssert(OSGiTest.groovy:181)
	at org.eclipse.smarthome.test.OSGiTest.waitForAssert(OSGiTest.groovy)
	at sun.reflect.GeneratedMethodAccessor20.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite$PogoCachedMethodSiteNoUnwrapNoCoerce.invoke(PogoMetaMethodSite.java:207)
	at org.codehaus.groovy.runtime.callsite.PogoMetaMethodSite.callCurrent(PogoMetaMethodSite.java:56)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCallCurrent(CallSiteArray.java:49)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:133)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.callCurrent(AbstractCallSite.java:141)
	at org.eclipse.smarthome.core.thing.internal.ThingManagerOSGiTest.ThingManager calls bridgeStatusChanged on ThingHandler correctly(ThingManagerOSGiTest.groovy:1057)
```

**Comment by sbussweiler** [comment]
@sjka This was a timing issue. I added some wait for assert statements...

@maggu2810 I'm not able to reproduce it. The default timeout (10 seconds) should be more than  enough. Please give me a hint if this error still occurs.

**Comment by sbussweiler** [comment]
I created the PR #2087. Please feel free to comment.

Please not that I created a new [branch](https://github.com/qivicon/smarthome/tree/thing-bridge-lifecycle).

**Comment by lolodomo** [comment]
Issue to close, no ?
